decimal64math.go

```diff
@@ -141,47 +141,40 @@ func (ctx Context64) Add(d, e Decimal64) Decimal64 {
 		}
 		return QNaN64
 	}
-	if dp.significand == 0 {
+	if dp.significand.lo == 0 {
 		return e
-	} else if ep.significand == 0 {
+	} else if ep.significand.lo == 0 {
 		return d
 	}
+	ep.removeZeros()
+	dp.removeZeros()
 	ep.updateMag()
 	dp.updateMag()
 	sep := dp.separation(ep)
-	roundStatus := matchScales(&dp, &ep)
-	// if the seperation of the numbers are more than 16 then we just return the larger number
-	if sep > 16 { // TODO: return rounded significand (for round down/Ceiling)
-		return d
-	} else if sep < -16 {
-		return e
+
+	if sep < 0 {
+		dp, ep = ep, dp
+		sep = -sep
 	}
-	var ans decParts
-	if ep.sign != dp.sign {
-		if ep.significand == dp.significand {
-			return Zero64
-		}
-		if dp.significand < ep.significand {
-			dp, ep = ep, dp
-		}
-		ans.sign = dp.sign
-		if dp.sign == 1 {
-			dp.sign, ep.sign = ep.sign, dp.sign
-		}
-	} else if dp.sign == 1 {
-		ans.sign = 1
-		dp.sign, ep.sign = 0, 0
-	} else {
-		ans.sign = 0
-	}
-	ans.significand = dp.significand + uint64(1-2*(ep.sign))*ep.significand
-	ans.exp = dp.exp
-	ans.exp, ans.significand = renormalize(ans.exp, ans.significand)
-	ans.significand = ctx.roundingMode.round(ans.significand, roundStatus)
-	if ans.exp > expMax || ans.significand > maxSig {
+	if sep > 17 {
+		return *dp.dec
+	}
+	var rndStatus discardedDigit
+	dp.matchScales128(&ep)
+	ans := dp.add128(&ep)
+	rndStatus = ans.roundToLo()
+	ans.updateMag()
+	if ans.exp < -expOffset {
+		rndStatus = ans.rescale(-expOffset)
+	}
+	ans.significand.lo = ctx.roundingMode.round(ans.significand.lo, rndStatus)
+	if ans.exp >= -expOffset && ans.significand.lo != 0 {
+		ans.exp, ans.significand.lo = renormalize(ans.exp, ans.significand.lo)
+	}
+	if ans.exp > expMax || ans.significand.lo > maxSig {
 		return infinities[ans.sign]
 	}
-	return newFromParts(ans.sign, ans.exp, ans.significand)
+	return newFromParts(ans.sign, ans.exp, ans.significand.lo)
```

Comment:
Instead of having to provide multiple paremeters through to newFromParts(), a new function that takes in the decpart will be cleaner, e.g. `ans.toParts()`